src/main/scala/is/hail/distributedmatrix/BlockMatrixIsDistributedMatrix.scala

```diff
@@ -0,0 +1,243 @@
+package is.hail.distributedmatrix
+
+import is.hail.utils._
+import org.apache.spark._
+import org.apache.spark.rdd.RDD
+import org.apache.spark.mllib.linalg._
+import org.apache.spark.mllib.linalg.distributed._
+
+import scala.reflect.ClassTag
+
+object BlockMatrixIsDistributedMatrix extends DistributedMatrix[BlockMatrix] {
+  type M = BlockMatrix
+
+  def cache(m: M): M = m.cache()
+
+  def from(irm: IndexedRowMatrix, dense: Boolean = true): M =
+    if (dense) irm.toBlockMatrixDense()
+    else irm.toBlockMatrix()
+  def from(bm: BlockMatrix): M = bm
+  def from(sc: SparkContext, dm: DenseMatrix, rowsPerBlock: Int, colsPerBlock: Int): M = {
+    val rbc = sc.broadcast(dm)
+    val rowBlocks = (dm.numRows - 1) / rowsPerBlock + 1
+    val colBlocks = (dm.numCols - 1) / colsPerBlock + 1
+    val rowsRemainder = dm.numRows % rowsPerBlock
+    val colsRemainder = dm.numCols % colsPerBlock
+    val indices = for {
+      i <- 0 until rowBlocks
+      j <- 0 until colBlocks
+    } yield (i, j)
+    val rMats = sc.parallelize(indices).map { case (i, j) =>
+      val rowsInThisBlock = (if (i + 1 == rowBlocks && rowsRemainder != 0) rowsRemainder else rowsPerBlock)
+      val colsInThisBlock = (if (i + 1 == colBlocks && colsRemainder != 0) colsRemainder else colsPerBlock)
+      val a = new Array[Double](rowsInThisBlock * colsInThisBlock)
+      for {
+        ii <- 0 until rowsInThisBlock
+        jj <- 0 until colsInThisBlock
+      } {
+        a(jj*rowsInThisBlock + ii) = rbc.value(i*rowsPerBlock + ii, j*colsPerBlock + jj)
+      }
+      ((i, j), new DenseMatrix(rowsInThisBlock, colsInThisBlock, a, false): Matrix)
+    }.partitionBy(GridPartitioner(rowBlocks, colBlocks))
+    new BlockMatrix(rMats, rowsPerBlock, colsPerBlock, dm.numRows, dm.numCols)
+  }
+
+  def transpose(m: M): M = m.transpose
+  def diagonal(m: M): Array[Double] = {
+    val rowsPerBlock = m.rowsPerBlock
+    val colsPerBlock = m.colsPerBlock
+
+    // FIXME: generalize, fix inner diagonal method
+    require(rowsPerBlock == colsPerBlock)
+
+    // check if the intervals [i, i+rows) and [j, j+rows) overlap,
+    // meaning this block contains the diagonal.
+    def containsDiagonal(coord: (Int, Int)): Boolean = coord match {
+      case (blocki, blockj) =>
+        val i = blocki * rowsPerBlock
+        val j = blockj * colsPerBlock
+        val i2 = i + rowsPerBlock
+        val j2 = j + colsPerBlock
+
+        i <= j && j < i2 ||
+        i < j2 && j2 < i2 ||
+        j <= i && i < j2 ||
+        j < i2 && i2 < j2
+    }
+
+    def diagonal(block: Matrix): Array[Double] = {
+      val length = math.min(block.numRows,block.numCols)
+      val diagonal = new Array[Double](length)
+      var i = 0
+      while (i < length) {
+        diagonal(i) = block(i, i)
+        i += 1
+      }
+      diagonal
+    }
+
+    m.blocks
+      .filter(x => containsDiagonal(x._1))
```

Comment:
under the assumption rowsPerBlock == colsPerBlock, this just becomes i == j, right?  I'd kill containsDiagonal and the max below, much easier to reason about that way.